app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala

```diff
@@ -476,8 +510,48 @@ class SectionRenderingService(
         htmlForFileUpload(fieldValue, formTemplateId, index, ei, data, userDetails, maybeValidated, lang)
       case InformationMessage(infoType, infoText) =>
         htmlForInformationMessage(fieldValue, infoType, infoText, index, ei)
+      case HmrcTaxPeriod(idType, idNumber, regimeType) =>
+        htmlForHmrcTaxPeriod(
+          fieldValue,
+          index,
+          ei,
+          maybeValidated,
+          data,
+          obligations,
+          HmrcTaxPeriod(idType, idNumber, regimeType))
     }
 
+  private def htmlForHmrcTaxPeriod(
+    fieldValue: FormComponent,
+    index: Int,
+    ei: ExtraInfo,
+    validatedType: ValidatedType,
+    data: FormDataRecalculated,
+    obligations: Map[HmrcTaxPeriod, TaxPeriods],
+    HmrcTP: HmrcTaxPeriod) = {
+
+    val taxPeriodList = obligations
+      .get(HmrcTP) match {
+      case Some(c) => c.taxPeriods
+      case _       => List[TaxPeriod]()
+    }
+
+    val taxPeriodOptions = taxPeriodList
+      .map(i => (formatDate(i.inboundCorrespondenceFromDate), formatDate(i.inboundCorrespondenceToDate), i.periodKey))
+      .map(i => new OptionParams(i._3, i._1, i._2, false))
```

Comment:
Accessing tuple element by index is not readable. Better approach would be to use

```
.map{ case (fromDate, toDate, value) => new OptionParams(value, fromDate, toDate, false)}
```

But why bother, we can construct `OptionParams` in previous call to `map`.